### ISSUE TITLE
Looping through token lists several items at a time

### ISSUE THREAD

**Issue by blefloch** [latex3/latex3#915]
On a regular basis I end up needing to go through a token list two items at a time and applying a two-argument function to them.  It would be useful to have `\tl_maptwo_inline:nn { a {bcd} {ef} g } { [#1/#2] }` give `[a/bcd][ef/g]`.  Of course, we shouldn't provide something so specific, but I think there is room for a more general `\tl_map_inline:npn { a {bcd} {ef} g } #1#2 { [#1/#2] }`.

Specifically, `\tl_map_inline:npnTF` (token list) (parameter text) `{` (replacement) `}` would define a macro as for `\cs_set:Npn` (macro) (parameter text) `{` (replacement) `}`, then repeatedly call this (macro) on the token list until it fails to fill the parameter text.  If nothing remained in the token list then the true branch would be taken and otherwise the false branch.  Examples:
```
\tl_map_inline:npnTF { a {bcd} {ef} g h } #1#2 { [#1/#2] } t f % => [a/bcd][ef/g]f
\tl_map_inline:npnTF { a , bc . , . d , c b a } #1 , #2 . { [#1/#2] } t f % => [a/bc][/]f
\tl_map_inline:npnTF { a , bc . , . d , c b a . } #1 , #2 . { [#1/#2] } t f % => [a/bc][/][d/cba]t
```
This is inspired by #819 but different (and slightly harder to implement), because the idea there was to do a standard map over items of the token list but then pass the item unbraced to some more complicated macro; this was straightforward to emulate with a standard `\tl_map_inline:nn` and an auxiliary function.

**Comment by car222222** [comment]
@blefloch If you need this then you are probably using the wrong data structure in your model/code.  Try simply using pairs as the items in your tl.

**Comment by car222222** [comment]
So what is needed are Creator/mutator commands to produce a tl with suitable groups. Similarly maybe for the other list data-types.

**Comment by blefloch** [comment]
The main use case is not for internal manipulation of data, but rather for "user" interface.  For instance we have
```
\int_case:nn { 123 }
  {
    { 0 } { code-for-0 }
    { 34 } { code-for-34 }
  }
```
in which it would not be good to ask the user to include extra braces around each {value}{code} pair.  Someone could want to implement an `\intrange_case:nn { value } { {min_1}{max_1}{code_1}   {min_2}{max_2}{code_2} ... }` that requires looping through a token list three items at a time.

I'm not particularly pushing for this, to be honest, let me mark this "for-later".

**Comment by car222222** [comment]
Agreed:   it would not be good to ask the user to include extra braces around each {value}{code} pair

But the internal data structure created by this inpu (and subsequently iterated over) should, at least in theory:-), be a ‘list of pairs’.   If this list is implemented using a tl data structure, then these pairs should be the items in the tl.  

Similarly for the three-tuples example: first get the data format correct, then look at simple interfaces for inputting the data, and then at processing it.

**Comment by blefloch** [comment]
Sure, internally you may want to turn the input into a list of pairs `{ {val_1}{code_1} } ...... { {val_n}{code_n} }`, but how do you transform the user input into that?  Well, you need to loop through the user's input, hence this proposed addition.

**Comment by car222222** [comment]
One might hope that there is a better way!  A proper constructor function for lists-of-tuples, for example.

But you are probably right that many somewhat arbitrary and general such low-level manipulations are needed (for tl lists at least).  But maybe they need a separate module (which could also contain other doubtful activities such as forgetful functors?).

I can draw up a list of such if you like:-).
